File: Day 4/0 Dynamic Array/solution.js
class DynamicArray {
    constructor() {
      this.data = new Array(1);
      this.size = 0;
    }
  
    get length() {
      return this.size;
    }
  
    get(index) {
      return this.data[index];
    }
  
    push(item) {
      if (this.size === this.data.length) {
        this._resize(2 * this.data.length);
      }
      this.data[this.size] = item;
      this.size++;
    }
  
    _resize(capacity) {
      const newData = new Array(capacity);
      for (let i = 0; i < this.size; i++) {
        newData[i] = this.data[i];
      }
      this.data = newData;
    }
  }
  

File: Day 4/0 Dynamic Array/solution.cs



public class DynamicArray<T>
{
    private T[] data;
    private int size;
    public int sizeFactor;

    public DynamicArray(int sizeFactor)
    {
        data = new T[10];
        size = 0;

        if(sizeFactor < 1.5){
                return NotSyupportedExcetopn();
        }

        sizeFactor  = sizeFactor;
    }


    public int Length
    {
        get { return size; }
    }

    public T this[int index]
    {
        get { return data[index]; }
        set { data[index] = value; }
    }

    public void Add(T item)
    {
        if (size == data.Length)
        {
            Resize(sizeFactor * data.Length);
        }
        
        data[size] = item;

        size++;
    }

    private void Resize(int capacity)
    {
        T[] newData = new T[capacity];
        for (int i = 0; i < size; i++)
        {
            newData[i] = data[i];
        }
        data = newData;
    }
}
